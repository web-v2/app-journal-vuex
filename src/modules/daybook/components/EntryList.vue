<template>
    <div>
        <div class="entry-list-container">
            <div class="px-2 pt-2">
                <input type="text" class="form-control" placeholder="Buscar entrada">
            </div>
            <div class="entry-scrollarea">
                <span>
                    <Entry v-for="items in 25" :key="items" />
                </span>
            </div>
        </div>
    </div>
</template>

<script>
import { defineAsyncComponent } from 'vue'
export default {
    components:{
        Entry: defineAsyncComponent(()=>import('./Entry.vue'))
    }
}
</script>

<style lang="scss" scoped>
.entry-list-container{
    border-right: 1px solid #2c3e50;
    height: calc(100vh - 56px);
}
.entry-scrollarea{
    height: calc(100vh - 110px);
    overflow: scroll;
}
</style>

