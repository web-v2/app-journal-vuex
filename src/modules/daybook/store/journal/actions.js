import journalApi from "@/api/journalAPI"
/*
export const myAction = async ({ commit }) => {

}
https://vue-demos-v2-default-rtdb.firebaseio.com/entries.json
*/

export const loadEntries = async ({ commit }) => {
    const {data} = await journalApi.get('/entries.json')
    if(!data){
        commit('setEntries', [])
        return
    }
    const entries = []
    for (let id of Object.keys(data)) {
        entries.push({
            id,
            ...data[id]
        })
    }
    commit('setEntries', entries)
}

export const updateEntry = async ({ commit }, entrada) => {
    const {fecha, picture, text} = entrada
    const dataToSave = {fecha, picture, text}
    const resp = await journalApi.put(`/entries/${entrada.id}.json`, dataToSave)    
    commit('updateEntry', {...entrada}) //commit updateEntry en Mutaciones para actualizar el state.
}

export const createEntry = async ({ commit }, entrada) => {
    const {fecha, picture, text} = entrada
    const dataToSave = {fecha, picture, text}
    //console.log(dataToSave);
    const {data} = await journalApi.post(`/entries.json`, dataToSave)    
    dataToSave.id = data.name;
    commit('addEntry', dataToSave)
    return data.name;
}

export const deleteEntry = async ({ commit }, id) => {
    const resp = await journalApi.delete(`/entries/${id}.json`)    
    commit('deleteEntry', id) //commit deleteEntry en Mutaciones para actualizar el state.
    return resp
}