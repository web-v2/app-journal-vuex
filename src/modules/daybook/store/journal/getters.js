/*
export const myGetter = ( state ) => {

    return state

}
*/

export const getEntriesByTerm = ( state ) => ( term = '') =>{
    if ( term.length === 0 ) return state.entries

    return state.entries.filter( entry => entry.text.toLowerCase().includes( term.toLocaleLowerCase() ))
}

export const getEntryById = ( state ) => (id = '') => {
    const ent = state.entries.find( entry => entry.id === id );
    if(!ent) return
    
    return { ...ent }
}