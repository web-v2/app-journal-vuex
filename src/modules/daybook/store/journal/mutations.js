/*
export const myMutation = ( state ) => {

}
*/
export const setEntries = ( state, entries ) => {
    state.entries = [...state.entries, ...entries];
    state.isLoading = false;
}

export const updateEntry = ( state, entry ) => {
    const idx = state.entries.map(e => e.id).indexOf(entry.id);
    state.entries[idx] = entry;
}

export const addEntry = (state, entry ) => {
    state.entries = [entry, ...state.entries];
}

export const deleteEntry = (state, id ) => {
    state.entries = state.entries.filter(e => e.id != id);
}