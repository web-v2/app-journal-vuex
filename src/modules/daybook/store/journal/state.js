//https://vue-demos-v2-default-rtdb.firebaseio.com/entries.json
export default () =>({
    isLoading: true,
    entries: []
})