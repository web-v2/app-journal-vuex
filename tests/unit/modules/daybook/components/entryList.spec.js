import { shallowMount } from "@vue/test-utils";
import { createStore } from "vuex";

import journal from "@/modules/daybook/store/journal";
import EntryList from "@/modules/daybook/components/EntryList.vue";

import { journalState } from "../../../mock-data/test-journal-state";

const createVuexStore = (initialState) =>
  createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...initialState },
      },
    },
  });

describe("Pruebas en el EntryList", () => {
  const store = createVuexStore(journalState);
  const mockRouter = {
    push: jest.fn(),
  };

  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryList, {
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    });
  });

  test("debe de llamar el getEntriesByTerm sin termino y mostrar 4 entradas", () => {
    expect(wrapper.findAll("entry-stub").length).toBe(4);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debe de llamar el getEntriesByTerm y filtrar las entradas", async () => {
    const input = wrapper.find("input");
    const resp = await input.setValue("Actualizando registro de una imgen");
    //const rtx = store.getters["journal/getEntriesByTerm"]("nueva entrada");
    //console.log({ rtx });

    expect(wrapper.findAll("entry-stub").length).toBe(1);
  });

  test("el boton de nuevo debe de redireccionar a /new", () => {
    wrapper.find("button").trigger("click");

    expect(mockRouter.push).toHaveBeenCalledWith({
      name: "entry",
      params: { id: "new" },
    });
  });
});
