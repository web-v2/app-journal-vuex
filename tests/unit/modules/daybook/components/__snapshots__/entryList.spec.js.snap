// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryList debe de llamar el getEntriesByTerm sin termino y mostrar 2 entradas 1`] = `
<div>
  <div class="entry-list-container">
    <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Buscar entrada"></div>
    <div class="mt-2 d-flex flex-column"><button class="btn btn-primary mx-3"><i class="fa fa-plus-circle"></i> Nueva entrada </button></div>
    <div class="entry-scrollarea"><span><entry-stub entry="[object Object]"></entry-stub><entry-stub entry="[object Object]"></entry-stub><entry-stub entry="[object Object]"></entry-stub><entry-stub entry="[object Object]"></entry-stub></span></div>
  </div>
</div>
`;

exports[`Pruebas en el EntryList debe de llamar el getEntriesByTerm sin termino y mostrar 4 entradas 1`] = `
<div>
  <div class="entry-list-container">
    <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Buscar entrada"></div>
    <div class="mt-2 d-flex flex-column"><button class="btn btn-primary mx-3"><i class="fa fa-plus-circle"></i> Nueva entrada </button></div>
    <div class="entry-scrollarea"><span><entry-stub entry="[object Object]"></entry-stub><entry-stub entry="[object Object]"></entry-stub><entry-stub entry="[object Object]"></entry-stub><entry-stub entry="[object Object]"></entry-stub></span></div>
  </div>
</div>
`;
